### PR TITLE
add the scheduling part of lifecycle

### PR DESCRIPTION
**Pull request by sfzeng**
Dataflow service will schedule according to lifecycle rules of each bucket which are configured by users, and send message to Datamover service through kafka. Datamover service will receive messages and do expiration action or transition action.
This PR only include the scheduling part.

### CODE REVIEW

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+func ScheduleLifecycle() {
+	log.Log("[ScheduleLifecycle] begin ...")
+	// Load transition map.
+	{
+		mutext.Lock()
+		defer mutext.Unlock()
+		if len(TransitionMap) == 0 {
+			err := loadStorageClassDefinition()
+			if err != nil {
+				log.Logf("[ScheduleLifecycle]Load storage classes failed: %v.\n", err)
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: s/Load/load

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+				return
+			}
+		}
+	}
+
+	// Get bucket list.
+	listReq := s3.BaseRequest{Id: "test"}
+	listRsp, err := s3client.ListBuckets(context.Background(), &listReq)
+	if err != nil {
+		log.Logf("[ScheduleLifecycle]List buckets failed: %v.\n", err)
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: s/List/list

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		return
+	}
+
+	for _, v := range listRsp.Buckets {
+		// For each bucket, get the lifecycle rules, and schedule each rule.
+		if v.LifecycleConfiguration == nil {
+			log.Logf("[ScheduleLifecycle]Bucket[%s] has no lifecycle rule.\n", v.Name)
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Same here.

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+			continue
+		}
+
+		log.Logf("[ScheduleLifecycle]Bucket[%s] has lifecycle rule.\n", v.Name)
+
+		err := handleBucketLifecyle(v.Name, v.Backend, v.LifecycleConfiguration)
+		if err != nil {
+			log.Logf("[ScheduleLifecycle]Handle bucket lifecycle for bucket[%s] failed, err:%v.\n", v.Name, err)
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Same here

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+func handleBucketLifecyle(bucket string, defaultBackend string, rules []*osdss3.LifecycleRule) error {
+	// Translate rules set by user to internal rules which can be sorted.
+
+	// Lifecycle scheduling must be mutual excluded among several schedulers, so get lock first.
+	ret := db.DbAdapter.LockBucketLifecycleSched(bucket)
+	for i := 0; i < 3; i++ {
+		if ret == LockSuccess {
+			break
+		} else if ret == LockBusy {
+			return fmt.Errorf("Lifecycle scheduleing of bucket[%s] is in progress.")
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: s/Lifecycle/lifecycle

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+		} else {
+			// Try to lock again, try three times at most
+			ret = db.DbAdapter.LockBucketLifecycleSched(bucket)
+		}
+	}
+	if ret != LockSuccess {
+		log.Logf("Lock scheduling failed.\n")
+		return fmt.Errorf("Internal error: Lock scheduling failed.")
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Same here

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+		log.Logf("abort rule: %+v\n", *v)
+	}
+	// End: Log for debug
+	schedSortedAbortRules(&abortRules)
+
+	return nil
+}
+
+func checkTransitionValidation(source int32, destination int32) bool {
+	valid := true
```
Comment (leonwanghui) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: There is no need to define valid variable here, just return true or false.

File: s3/pkg/service.go [hunk 8]
```diff
@@ -538,3 +538,36 @@ func (b *s3Service) GetBackendTypeByTier(ctx context.Context, in *pb.GetBackendT
 
 	return nil
 }
+func (b *s3Service) AddUploadRecord(ctx context.Context, record *pb.MultipartUploadRecord, out *pb.BaseResponse) error {
+	log.Logf("add multipart upload record")
+	err := db.DbAdapter.AddMultipartUpload(record)
+	if err != NoError {
+		return err.Error()
```
Comment (leonwanghui) [inline s3/pkg/service.go]: what do u mean by `err.Error()`

File: s3/pkg/service.go [hunk 9]
```diff
@@ -538,3 +538,36 @@ func (b *s3Service) GetBackendTypeByTier(ctx context.Context, in *pb.GetBackendT
...
+	}
+
+	return nil
+}
+
+func (b *s3Service) DeleteUploadRecord(ctx context.Context, record *pb.MultipartUploadRecord, out *pb.BaseResponse) error {
+	log.Logf("delete multipart upload record")
+	err := db.DbAdapter.DeleteMultipartUpload(record)
+	if err != NoError {
+		return err.Error()
```
Comment (leonwanghui) [inline s3/pkg/service.go]: Same here.

File: s3/pkg/service.go (hunk 8 above)
Comment (sfzeng) [reply s3/pkg/service.go]: return value type of DbAdapter S3Error not errors, but we need return errors not S3Error

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 1 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 2 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 3 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 4 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 5 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 6 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go (hunk 7 above)
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: done

**Pull request merged**